src/main/java/io/strimzi/kafka/quotas/VolumeSourceBuilder.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.kafka.quotas;
+
+import java.util.Collection;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import org.apache.kafka.clients.admin.Admin;
+import org.apache.kafka.clients.admin.AdminClient;
+import org.apache.kafka.clients.admin.LogDirDescription;
+
+/**
+ * A builder which ensures the <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-827%3A+Expose+log+dirs+total+and+usable+space+via+Kafka+API">KIP-827 API</a> is available and will throw exceptions if not.
+ *
+ * @see <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-827%3A+Expose+log+dirs+total+and+usable+space+via+Kafka+API">KIP-827 API</a>
+ */
+public class VolumeSourceBuilder implements AutoCloseable {
+
+    private final Supplier<Boolean> kip827Available;
+    private final Function<StaticQuotaConfig.KafkaClientConfig, Admin> adminClientFactory;
+    private Admin adminClient;
+    private StaticQuotaConfig config;
+    private Consumer<Collection<Volume>> volumesConsumer;
+
+    /**
+     * Default production constructor for production usage.
+     * Which will lazily create a Kafka admin client using the supplied config.
+     */
+    @SuppressFBWarnings("MC_OVERRIDABLE_METHOD_CALL_IN_CONSTRUCTOR") //false positive we are just passing the method reference
+    public VolumeSourceBuilder() {
+        this(VolumeSourceBuilder::testForKip827, kafkaClientConfig -> AdminClient.create(kafkaClientConfig.getKafkaClientConfig()));
+    }
+
+    /**
+     * Secondary constructor visible for testing.
+     * @param kip827Available used to determine if KIP-827 API's are available
+     * @param adminClientFactory factory function for creating Admin clients with the builders' config.
+     */
+    /* test */ VolumeSourceBuilder(Supplier<Boolean> kip827Available, Function<StaticQuotaConfig.KafkaClientConfig, Admin> adminClientFactory) {
+        this.kip827Available = kip827Available;
+        this.adminClientFactory = adminClientFactory;
+    }
+
+     /*test*/ static Boolean testForKip827() {
```

Comment:
We kept this in as a warning to anyone who tries to use this code with an old broker, even though we know for Strimzi it should only be used with compatible Kafka versions. Is this reasonable, or should we drop this code and put some warning in the README about compatibility?

Reply:
I was originally planning on making the migration story a bit cleaner, I think its worth keeping the check for now as it is a hard requirement for the plug-in